src/util/auth.ts

```diff
@@ -0,0 +1,92 @@
+import { authConfig, basePaths } from "../config/auth";
+
+export const auth = {
+    parseUri: (uri: string): { route: string; id: string } => {
+        // load list of routes
+        const routes: string[] = Object.keys(authConfig);
+        // remove query params
+        uri = uri.split("?")[0];
+        // remove extra "/" at the end if it's there
+        if (uri[uri.length - 1] == "/") uri = uri.substr(0, uri.length - 1);
+
+        // check if there is a route matching the uri
+        // (saves time if the route doesn't have route params)
+        if (routes.includes(uri.toLowerCase()))
+            return { route: uri.toLowerCase(), id: "" };
+        else {
+            // split uri and remove 1st empty position
+            const uriSlice = uri.split("/");
+            uriSlice.shift();
+            // extract id (can be any resource id but only user
+            // ids will be used in case of a user-protected route)
+            const id = uriSlice.length > 1 ? uriSlice[1] : "";
+
+            // lowercase all array elements
+            for (let i = 0; i < uriSlice.length; i++)
+                uriSlice[i] = uriSlice[i].toLowerCase();
+
+            // get base path
+            const basePath = uriSlice[0];
+
+            // if basepath doesn't exist, then route doesn't exist
+            if (!basePaths.includes(basePath)) return { route: "", id: "" };
+            else {
+                // remove the id variable to something consistent
+                if (id) uriSlice[1] = "{id}";
+                return {
+                    route: "/" + uriSlice.join("/"),
+                    id: id ? id : ""
+                };
+            }
+        }
+    },
+
+    authorize: (
+        route: string,
+        method: string,
+        role: string,
+        requestUserId: string,
+        resourceUserId: string
```

Comment:
it will be used by other resources but this id will only be used in the case of the user microservice (we compare the id of the user that issued the request to the id of the user to whom the resource belongs)

Reply:
So is the plan to have a different `authorize` method for each microservice?

Reply:
no. if a resource belongs to a user, it will reference which user it belongs to

Reply:
Okay maybe I'm not understanding: what about in the future case of the CMS microservice with something like 
```/blogs/{blogId}```?
We don't have the resource's user ID from the URI, so we'll have to have some special logic to fetch the blog from mongo and see which user it belongs to. Where are we planning on having this logic?

This is definitely in the future, but I'm trying to figure out a design that works for resources that belong to users but those resources aren't managed by the user microservice

Reply:
in those cases, we'll have to expand on our current authorize logic but it isn't something that we need to do until we get cases that require it

Reply:
none of the future changes would break our current logic so we can leave it for later

Reply:
Okay, sounds good 👌 we can tackle that when we get there